modules/MatchProvider.js

```diff
@@ -17,34 +18,35 @@ class MatchProvider extends React.Component {
     // **IMPORTANT** we must mutate matches, never reassign, in order for
     // server rendering to work w/ the two-pass render approach for Miss
     this.matches = []
```

Comment:
If you add the `'doesn't render on first pass'` test to the current `v4` branch, it should fail.

If you look at the `render` function of `<Miss>`, it determines if it should render by calling the `missedAtIndex` function. The current `missedAtIndex` code just checks if the current item (determined by `serverRouterIndex`) has matches in the `matchesByIdentity` array and if there is a `<Miss>` component for it.

The second check will always be true because the `<Miss>` will have registered itself. The first check will be true (aka there are no matches) if no `<Match>` components match **prior** to the `<Miss>` component rendering. If there is `<Match>` that matches the current location, but it is rendered after a `<Miss>`, that `<Miss>` will be rendered on the first pass.